python/cuml/experimental/explainer/kernel_shap.pyx

```diff
@@ -0,0 +1,505 @@
+            Bw = y_hat * cp.sqrt(self.weights)
+            X, *_ = cp.linalg.lstsq(Aw, Bw)
+
+            return X
+        else:
+            y_hat = y_hat[nonzero_inds] - self.expected_value
+            Aw = self.mask[nonzero_inds] * cp.sqrt(
+                self.weights[nonzero_inds, cp.newaxis]
+            )
+            Bw = y_hat * cp.sqrt(self.weights[nonzero_inds])
+            X, *_ = cp.linalg.lstsq(Aw, Bw)
+
+            return X
+
+    def shap_values(self, X, l1_reg='auto'):
+        """
+        Legacy interface to estimate the SHAP values for a set of samples.
+
+        Parameters
+        ----------
+        X : Dense matrix containing floats or doubles.
+            Acceptable formats: CUDA array interface compliant objects like
+            CuPy, cuDF DataFrame/Series, NumPy ndarray and Pandas
+            DataFrame/Series.
+        l1_reg : str (default: 'auto')
+            The l1 regularization to use for feature selection.
+
+        Returns
+        -------
+        array or list
+
+        """
+        return self.explain(X, l1_reg)
+
+    def __call__(self,
+                 X,
+                 l1_reg='auto'):
+        if has_shap():
+            warn("SHAP's Explanation object is still experimental, the main "
+                 "API currently is 'explainer.shap_values'.")
+            from shap import Explanation
+            res = self.explain(X, l1_reg)
+            out = Explanation(
+                values=res,
+                base_values=self.expected_value,
+                base_values=self.expected_value,
+                data=self.background,
+                feature_names=self.feature_names,
+            )
+            return out
+        else:
+            raise ImportError("SHAP package required to build Explanation "
+                              "object. Use the explainer.shap_values "
+                              "function to get the shap values, or install "
+                              "SHAP to use the new API style.")
+
+
+@lru_cache(maxsize=None)
+def binomCoef(n, k):
+    res = 1
+    if(k > n - k):
+        k = n - k
+    for i in range(k):
+        res *= (n - i)
+        res /= (i + 1)
+
+    return res
+
+
+def powerset(n, r, nrows, dtype=np.float32):
+    N = np.arange(n)
+    w = np.zeros(nrows, dtype=dtype)
+    result = np.zeros((nrows, n), dtype=dtype)
+    idx = 0
+    for i in range(1, r + 1):
+        for c in combinations(N, i):
+            result[idx, c] = 1
+            w[idx] = shapley_kernel(n, i)
+            idx += 1
+
+    return result, w
+
+
+def calc_sample_weights(M, r):
+    w = np.empty(M - r, dtype=np.float32)
+    for i in range(M - r, M):
+        # w[i] = shapley_kernel(nsamples, i)
+        w[i] = (M - 1) / i * (M - i)
+    return w
+
+
+@lru_cache(maxsize=None)
+def shapley_kernel(M, s):
+    if(s == 0 or s == M):
+        return 10000
+
+    res = (M - 1) / (binomCoef(M, s) * s * (M - s))
+    return res
```

Comment:
If these functions arent used outside this file, make them private with a leading underscore.